### PR TITLE
TorchX-MCAD support

### PR DESCRIPTION
**Pull request by Sara-KS**
Add support to TorchX for the MCAD Kubernetes Scheduler

Test plan:
Created a Kubernetes MCAD scheduler test and added it to the TorchX test suite.

### CODE REVIEW

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,1121 @@
...
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+
+This contains the TorchX Kubernetes scheduler which can be used to run TorchX
+components on a Kubernetes cluster.
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:11]: nit: update to refer to MCAD

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,1121 @@
...
+Prerequisites
+==============
+
+TorchX kubernetes scheduler depends on AppWrapper + MCAD and requires etcd intalled for distributed job execution.
+
+Install MCAD 
+
+.. Deploying Multi-Cluster-Application-Dispatcher
+
+    https://github.com/project-codeflare/multi-cluster-app-dispatcher/blob/main/doc/deploy/deployment.md
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:22]: just make this a link instead of a code block

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,1121 @@
...
+
+
+def sanitize_for_serialization(obj: object) -> object:
+    from kubernetes import client
+
+    api = client.ApiClient()
+    return api.sanitize_for_serialization(obj)
+
+
+def role_to_pod(
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:173]: Is it possible to share this method between the kubernetes scheduler and mcad? I see the env differences, anything else?

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,1121 @@
...
+    pattern = r"[a-z0-9\-]"
+    return "".join(re.findall(pattern, data.lower()))
+
+
+def get_port_for_service(app: AppDef) -> str:
+    # Initialize port to default
+    port = "29500"
+
+    # look for rdzv_endpoint
+    rdzv = "--rdzv_endpoint"
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:431]: this seems pretty hacky -- can we instead grab the ports from the port_map in the role?

https://github.com/pytorch/torchx/blob/main/torchx/components/dist.py#L203

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,1121 @@
...
+            genericitem: Dict[str, Any] = {
+                "replicas": 1,
+                "generictemplate": pod,
+            }
+            # TODO: retry support here
+            if role.max_retries > 0:
+                genericitem["maxRetry"] = role.max_retries
+                genericitem["policies"] = RETRY_POLICIES[role.retry_policy]
+                msg = f"""
+Role {role.name} configured with restarts: {role.max_retries}.
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:500]: nit: add comment about why we're warning about this "MCAD doesn't support retries. Role ... configured..."

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,1121 @@
...
+    ]
+
+    my_env_var = [
+        V1EnvVar(
+            name=f"MCAD_{cleanup_str(role.name)}_{replica_id}_HOSTS".upper().replace(
+                "-", ""
+            ),
+            value=f"{unique_app_id}-{replica_id}.{unique_app_id}",
+        )
+        for replica_id in range(role.num_replicas)
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:317]: nit: these environment variables aren't strictly needed, just one for the rank 0 host

may want to prefix it with TORCHX_MCAD so users don't think they're provided by the scheduler

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,1121 @@
...
+    """
+
+    service_port = get_port_for_service(app)
+
+    if not (0 < int(service_port) <= 65535):
+        msg = """Warning: rdzv_endpoint set to invalid port number. Value must be between 1-65535, with torchx default = 29500. Try updating --rdzv_port argument to dist.ddp. Setting port to default = 29500"""
+        service_port = "29500"
+        warnings.warn(msg)
+
+    svc_obj = mcad_svc(unique_app_id, namespace, service_port)
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:519]: what's this service used for? for pytorch rendezvous it's using the hostnames and not this right?

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,1121 @@
...
+    mounts_key = "volumeMounts"
+
+    # resource keys
+    cpu_key = "cpu"
+    gpu_key = "gpu"
+    request_key = "requests"
+    mem_key = "memory"
+
+    for genericItem in genericItems:
+        if gt_key in genericItem.keys():
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:574]: this is really nested, can we flatten a lot of these out with continues?

i.e.

```
if gt_key not in genericItem.keys():
   continue
gt_result = generic_item[gt_key]
...
```

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,1121 @@
...
+
+
+# Helper functions for MCAD generic items information -> TorchX Role
+def get_role_information(genericItems: Iterable[Dict[str, Any]]) -> Dict[str, Any]:
+    # Store unique role information
+    roles = {}
+
+    # nested dictionary keys
+    # meta data information
+    gt_key = "generictemplate"
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:552]: nit: all caps for constants GT_KEY

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,1121 @@
...
+        "apiVersion": "mcad.ibm.com/v1beta1",
+        "kind": "AppWrapper",
+        "metadata": {"name": unique_app_id},
+        "spec": job_spec,
+    }
+    return resource
+
+
+# Helper functions for MCAD generic items information -> TorchX Role
+def get_role_information(genericItems: Iterable[Dict[str, Any]]) -> Dict[str, Any]:
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:546]: nit: snake case

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,1121 @@
...
+
+        namespace, name = app_id.split(":")
+
+        print("\033[100m" + "\n ks - log_iter \n" + "\033[0m")
+        print(f"{name}")
+
+        pod_name = cleanup_str(f"{name}-{k}")
+
+        print("\033[100m" + "\n ks - log_iter \n" + "\033[0m")
+        print(f"{pod_name}")
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:1056]: nit: remove debug prints

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,1121 @@
...
+            iterator = split_lines(resp)
+
+        if regex:
+            return filter_regex(regex, iterator)
+        else:
+            return iterator
+
+    def list(self) -> List[ListAppResponse]:
+        p1 = subprocess.run(
+            ["kubectl", "config", "view", "--minify"],
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:1081]: this is a clever trick :)

Might be better to grab it from the kube config using the Python API however

https://github.com/kubernetes-client/python/issues/363#issuecomment-389722355

Would be nice to integrate this into the normal k8s scheduler as well

**Review by d4l3k — COMMENTED** [review]
It's looking good! Excited to see this :)

Would be nice to share some of these helper methods/spec generation with the Volcano integration where possible but if it doesn't make sense that's fine

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 13]
```diff
@@ -0,0 +1,1121 @@
...
+    ReplicaStatus,
+    Resource,
+    RetryPolicy,
+    Role,
+    RoleStatus,
+    runopts,
+    VolumeMount,
+)
+
+# from torchx.workspace.docker_workspace import DockerWorkspace
```
[the remark this answers is not in the document]
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Removed in commit: 65362828973a37624a6774a6da36a7460b4f6a89

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 14]
```diff
@@ -0,0 +1,1121 @@
...
+
+TorchX uses `torch.distributed.run <https://pytorch.org/docs/stable/elastic/run.html>`_ to run distributed training.
+This requires the installation of etcd service on your kubernetes cluster:
+
+.. code:: bash
+
+    kubectl apply -f https://github.com/pytorch/torchx/blob/main/resources/etcd.yaml
+
+
+Learn more about running distributed trainers :py:mod:`torchx.components.dist`
```
[the remark this answers is not in the document]
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Removed in commit: 65362828973a37624a6774a6da36a7460b4f6a89

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 1 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated in commit: 65362828973a37624a6774a6da36a7460b4f6a89

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 2 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated in commit: 65362828973a37624a6774a6da36a7460b4f6a89

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 3 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:165]: Primary differences are the function arguments (Volcano options or MCAD options) and the addition of the environment variable. This also leads to differences in the pod specs and metadata (section on V1Pod).

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 8 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Restructured in commit: 2c4a0c91493960d11a25068ab7729ac5eaf25177

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 9 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated constants in commit: 2c4a0c91493960d11a25068ab7729ac5eaf25177

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 10 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: updated in commit: 2c4a0c91493960d11a25068ab7729ac5eaf25177

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 7 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:511]: This service is used for distributed jobs to enable the PyTorch rendezvous. Since MCAD treats the jobs holistically, it needs to know the Kubernetes service information and pod information and treats them as the same "AppWrapped" job. For example, a distributed job that will run as two pods will be represented in a single "App wrapper" as a pod with 2 replicas and 1 service. This includes needing the port that will be used for the PyTorch rendezvous. The Kubernetes Volcano scheduler automatically creates and destroys a service for distributed PyTorch jobs as well. However, Volcano manages this differently and so TorchX does not need to be aware of the Kubernetes service.

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 4 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:423]: Distributed PyTorch jobs fail if the ports do not match, this is related to the conversation on the mcad_svc function below.

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 5 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated message in commit: bcb0ebe432db59f8ecca3c0ac77bd8bdc2d5a83a

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 6 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated prefix in commit: bcb0ebe432db59f8ecca3c0ac77bd8bdc2d5a83a

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 12 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated MCAD and Volcano schedulers to use Kube config for getting the namespace in both list functions. Commit: bcb0ebe432db59f8ecca3c0ac77bd8bdc2d5a83a

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 4 above)
Comment (d4l3k) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:421]: interesting -- can we still pull the port from the AppDef port map instead of introspecting like this? would be nice to support jobs that aren't using torch elastic

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 7 above)
Comment (d4l3k) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:509]: is the service set as the hostname on the workers? I'm just not quite understanding why a service is needed here if the pods can directly talk to each other.  The service is for load balancing/discovery but we're using the pod host names so if I understanding correctly shouldn't be needed

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 15]
```diff
@@ -0,0 +1,1090 @@
...
+
+"""
+
+This contains the TorchX Kubernetes_MCAD scheduler which can be used to run TorchX
+components on a Kubernetes cluster via the Multi-Cluster-Application-Dispatcher (MCAD).
+
+Prerequisites
+==============
+
+TorchX kubernetes scheduler depends on AppWrapper + MCAD..
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:16]: nit: extra period

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 4 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Reran tests and was able to update to use the role port_map. Commit: 697c5aa0b348cb38b6c79447aa70eda57f5fcbcb

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 7 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:501]: This section just generates the yaml specs for a Kubernetes service. TorchX-Volcano generates a similar Kubernetes Service as well, but because Volcano handles this automation the rest of TorchX does not even need to be aware of the Kubernetes service. MCAD does not have this type of automation in place, so it had to be moved to the TorchX scheduler level and included in the generate yaml file that is evaluated by MCAD. If one were to simply comment out this mcad service and try to run a distributed job, the main worker stalls at waiting to rendezvous with all workers and the replica workers will reach a time out.

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 15 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: removed

**Review by d4l3k — APPROVED** [review]
LGTM, thanks for contributing!

Just waiting for CI to pass and then will merge

**Comment by d4l3k** [comment]
docsbuild error seems unrelated, landing
